h/h_api/exceptions.py

```diff
@@ -0,0 +1,67 @@
+from h.h_api.model.json_api import JSONAPIError, JSONAPIErrorBody
+
+
+class JSONAPIException(Exception):
+    """An exception which can turn itself into a a JSON API error response."""
+
+    http_status = None
+
+    def _error_bodies(self):
+        """Get the instances of JSONAPIErrorBody representing this error."""
+        raise NotImplementedError()
+
+    def as_dict(self):
+        """Get a JSON API compatible dict representing this error."""
+        return JSONAPIError.create(self._error_bodies()).raw
+
+
+class SchemaValidationError(JSONAPIException):
+    """Represent a number of schema validation errors.
+
+    This class can be used to build up the errors and then check if the
+    exception should be raised.
+    """
+
+    http_status = 400
+
+    def __init__(self, title=None):
+        if title is None:
+            title = "The provided data does not match the schema"
+
+        super().__init__(title)
+        self.error_bodies = []
+
+    def add_error(self, error):
+        """
+        Add an error to this exception.
+
+        :param error: A `jsonschema.exceptions.ValidationError` error
+        """
+        self.error_bodies.append(
+            JSONAPIErrorBody.create(
+                self,
+                detail=error.message,
+                pointer=self._path_to_string(error.path),
+                meta={
+                    "schema": {"pointer": self._path_to_string(error.schema_path)},
+                    "context": [context.message for context in error.context],
+                },
```

Comment:
There's a fantastic amount of information inside schema errors from `jsonschema`. We'll pick some highlights.